### PR TITLE
ingest dbt relationship tests as additional_deps

### PR DESCRIPTION
**Pull request by johannkm**
dbt relationship tests depend on multiple models. We can now store that dependency on the AssetCheckSpec. This doesn't change behavior at all currently, but it will enable
- showing the dependency in the ui
- warnings, e.g. are you sure you want to run this check? additional_dep xyz hasn't materialized

### CODE REVIEW

**Comment by johannkm** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #19319** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19319?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/19319?utm_source=stack-comment).

File: python_modules/libraries/dagster-dbt/dagster_dbt_tests/test_asset_checks.py [hunk 1]
```diff
@@ -73,6 +74,102 @@ def my_dbt_assets_with_checks():
         assert any(unique_id.startswith("test") for unique_id in manifest["nodes"].keys())
         assert asset_def.check_specs_by_output_name
 
+        assert asset_def.check_specs_by_output_name == {
```
Comment (johannkm) [inline python_modules/libraries/dagster-dbt/dagster_dbt_tests/test_asset_checks.py:77]: could put this in a snapshot? But I think it's small enough to be fine

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_utils.py [hunk 2]
```diff
@@ -530,25 +529,36 @@ def is_generic_test_on_attached_node_from_dbt_resource_props(
...
-            dagster_dbt_translator_settings.enable_asset_checks,
+            dagster_dbt_translator.settings.enable_asset_checks,
             is_generic_test_on_attached_node,
         ]
     ):
         return None
 
+    additional_deps = []
+
+    depends_on = dbt_resource_props.get("depends_on", {}).get("nodes", [])
```
Comment (johannkm) [inline python_modules/libraries/dagster-dbt/dagster_dbt/asset_utils.py:550]: need to run this through more dbt projects to test it

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_utils.py [hunk 3]
```diff
@@ -530,25 +529,39 @@ def is_generic_test_on_attached_node_from_dbt_resource_props(
...
-            dagster_dbt_translator_settings.enable_asset_checks,
+            dagster_dbt_translator.settings.enable_asset_checks,
             is_generic_test_on_attached_node,
         ]
     ):
         return None
 
+    test_unique_id = dbt_resource_props["unique_id"]
+    # tests can't depend on tests, but include "test" so that we collect deps for test_unique_id
+    check_deps = get_deps(
```
Comment (johannkm) [inline python_modules/libraries/dagster-dbt/dagster_dbt/asset_utils.py:550]: @rexledesma is this a reasonable use of get_deps? It's possible we don't need the ephemeral -> actual model resolution here

File: python_modules/dagster/dagster/_core/definitions/asset_check_spec.py [hunk 4]
```diff
@@ -98,6 +98,13 @@ def __new__(
             additional_deps, AssetCheckKey(asset_key, name)
         )
 
+        for dep in additional_asset_deps:
+            if dep.asset_key == asset_key:
+                raise ValueError(
+                    f"Asset check {name} for asset {asset_key.to_string()} cannot have an addtional dependency "
+                    f"on asset{asset_key.to_string()}."
```
Comment (rexledesma) [inline python_modules/dagster/dagster/_core/definitions/asset_check_spec.py:105]: Can we split out dagster core changes into a separate PR? Also, small typo here:

```suggestion
                    f"Asset check {name} for asset {asset_key.to_string()} cannot have an additional dependency "
                    f"on asset {asset_key.to_string()}."
```

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorator.py [hunk 5]
```diff
@@ -411,7 +411,11 @@ def get_dbt_multi_asset_args(
         for test_unique_id in test_unique_ids:
             test_resource_props = manifest["nodes"][test_unique_id]
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorator.py:412]: [1]

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_utils.py (hunk 3 above)
Comment (rexledesma) [reply python_modules/libraries/dagster-dbt/dagster_dbt/asset_utils.py:550]: Can we just figure this out using the `manifest` at [1]? This is essentially just all the extra models in `manifest["parent_map"][test_unique_id]`.

File: python_modules/dagster/dagster/_core/definitions/asset_check_spec.py (hunk 4 above)
Comment (johannkm) [reply python_modules/dagster/dagster/_core/definitions/asset_check_spec.py:105]: https://github.com/dagster-io/dagster/pull/19358

**Comment by johannkm** [comment]
switched to parent_map

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_utils.py [hunk 6]
```diff
@@ -530,25 +529,34 @@ def is_generic_test_on_attached_node_from_dbt_resource_props(
 def default_asset_check_fn(
     asset_key: AssetKey,
     unique_id: str,
-    dagster_dbt_translator_settings: "DagsterDbtTranslatorSettings",
+    dagster_dbt_translator: "DagsterDbtTranslator",
     dbt_resource_props: Mapping[str, Any],
+    dbt_nodes: Mapping[str, Any],
+    parent_ids: List[str],
 ) -> Optional[AssetCheckSpec]:
```
Comment (rexledesma) [inline python_modules/libraries/dagster-dbt/dagster_dbt/asset_utils.py:536]: Just use `manifest`, so we don't have to pass in two arguments. Refactor `default_asset_check_fn` to the following arguments:

- `manifest`,
- `dagster_dbt_translator`,
- `asset_key`,
- `unique_id`
- `test_unique_id`

Then all the necessary indexing into `manifest` can be done in the body of `default_asset_check_fn`.

**Review by rexledesma — APPROVED** [review]
see comment
